### PR TITLE
add user-agent based on environment variable - opt in

### PR DESCRIPTION
**Pull request by ScottSuarez**
Add an extension to the user agent if the an environment variable is specified.

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

File: mmv1/third_party/terraform/utils/header_transport.go [hunk 1]
```diff
@@ -20,6 +21,10 @@ func newTransportWithHeaders(baseTransit http.RoundTripper) headerTransportLayer
 	if requestReason := os.Getenv("CLOUDSDK_CORE_REQUEST_REASON"); requestReason != "" {
 		headers.Set("X-Goog-Request-Reason", requestReason)
 	}
+	// opt in extension for adding to the User-Agent header
```
Comment (rileykarson) [inline mmv1/third_party/terraform/utils/header_transport.go:24]: Can we set this in `config.go` instead so all of the user agent stuff is in one place?

**Review by rileykarson — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/utils/header_transport.go (hunk 1 above)
Comment (ScottSuarez) [reply mmv1/third_party/terraform/utils/header_transport.go]: done

File: mmv1/third_party/terraform/utils/config.go.erb [hunk 2]
```diff
@@ -173,6 +173,10 @@ func (c *Config) LoadAndValidate(ctx context.Context) error {
 	// 4. Header Transport - outer wrapper to inject additional headers we want to apply
 	// before making requests
 	headerTransport := newTransportWithHeaders(retryTransport)
+	// opt in extension for adding to the User-Agent header
+	if ext := os.Getenv("GOOGLE_TPG_USERAGENT_EXTENSION"); ext != "" {
+		headerTransport.Set("User-Agent", ext)
```
Comment (rileykarson) [inline mmv1/third_party/terraform/utils/config.go.erb:178]: Oh, sorry, I thought `providerConfigure` from provider.go was in config.go. Let's modify Config.userAgent there instead of adding new functionality to the transport.

File: mmv1/third_party/terraform/utils/header_transport.go [hunk 3]
```diff
@@ -29,6 +30,11 @@ func (h headerTransportLayer) RoundTrip(req *http.Request) (*http.Response, erro
 		// only set headers that are not previously defined
 		if _, ok := req.Header[key]; !ok {
 			req.Header[key] = value
+		} else if key == "User-Agent" {
```
Comment (rileykarson) [inline mmv1/third_party/terraform/utils/header_transport.go:33]: This should be obsoleted by just modifying Config.userAgent.

File: mmv1/third_party/terraform/utils/config.go.erb (hunk 2 above)
Comment (ScottSuarez) [reply mmv1/third_party/terraform/utils/config.go.erb]: done

File: mmv1/third_party/terraform/utils/header_transport.go (hunk 3 above)
Comment (ScottSuarez) [reply mmv1/third_party/terraform/utils/header_transport.go]: done

File: mmv1/third_party/terraform/utils/provider.go.erb [hunk 4]
```diff
@@ -496,6 +496,12 @@ func providerConfigure(ctx context.Context, d *schema.ResourceData, p *schema.Pr
         <% end -%>
 	}
 
+	// opt in extension for adding to the User-Agent header
+	if ext := os.Getenv("GOOGLE_TPG_USERAGENT_EXTENSION"); ext != "" {
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/utils/provider.go.erb:500]: @rileykarson is the env variable name okay?

`GOOGLE_TERRAFORM_PROVIDER_USERAGENT_EXTENSION` was too wordy
`TPG_USERAGENT_EXTENSION` was to ambiguous

Comment (rileykarson) [reply mmv1/third_party/terraform/utils/provider.go.erb:500]: wdyt of just `GOOGLE_TERRAFORM_USERAGENT_EXTENSION`? Since this'll apply across multiple providers.

Comment (ScottSuarez) [reply mmv1/third_party/terraform/utils/provider.go.erb]: i dig it

File: mmv1/third_party/terraform/utils/provider.go.erb [hunk 5]
```diff
@@ -496,6 +496,12 @@ func providerConfigure(ctx context.Context, d *schema.ResourceData, p *schema.Pr
         <% end -%>
 	}
 
+	// opt in extension for adding to the User-Agent header
+	if ext := os.Getenv("GOOGLE_TERRAFORM_USERAGENT_EXTENSION"); ext != "" {
+		ua := config.userAgent
+		config.userAgent = fmt.Sprintf("%s %s", ua, ext)
```
Comment (rileykarson) [inline mmv1/third_party/terraform/utils/provider.go.erb:502]: nit: you can just inline `config.userAgent` in for `ua`

**Review by rileykarson — APPROVED** [review]

File: mmv1/third_party/terraform/utils/provider.go.erb (hunk 5 above)
Comment (ScottSuarez) [reply mmv1/third_party/terraform/utils/provider.go.erb:502]: noop for readability imo
